README.md
# pythia

Pythonic Deepstream.

---
[![PyPI - Version](https://img.shields.io/pypi/v/pythiags)](https://pypi.org/project/pythiags/)
[![PyPI - Wheels](https://img.shields.io/pypi/wheel/pythiags)](https://pypi.org/project/pythiags/)
[![Docs](https://img.shields.io/badge/docs-github_pages-blue)](https://rmclabs-io.github.io/pythia-docs/)

---

[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![security: bandit](https://img.shields.io/badge/security-bandit-yellow.svg)](https://github.com/PyCQA/bandit)

---


[NVidia Deepstream](https://developer.nvidia.com/deepstream-sdk) is an excellent gstreamer
framework which allows to build ai-powered, performant applications running on nvidia
hardware. Its python API and bindings, however, have a bunch of painpoints which we've
here collected and addressed with pythia:

* **Metadata Extraction**: Deepstream metadata extraction requires using buffer probes:
  pythia provides an easy to use interface which splits metadata extraction and
  processing.
* **Metadata Iteration**: pyds api iterators are not pythonic: pythia provides intuitive 
  deepstream metadata iterators to use, like `for frame in frames_per_batch(buffer)`
  wrapping pybind c++ casting and iteration.
* **Python boilerplate**: Python gstreamer apps get very large very fast. Pythia abstracts
  away common stuff and lets you focus on your application.
* **Quick prototyping**: Sometimes you just want to check the performance of a new model
  (eg after exporting from [Nvidia TAO](https://developer.nvidia.com/tao)), or verify
  the environment. Pythia comes with ready-to-run demo applications, and a
  `gst-launch`-like cli.

pythia offers:

* Common metadata extraction and parsing utilities.
* Workers and queues management in the background, to offload processing outside of the 
  buffer probe.
* Ready to use Docker images for both aarch64 (jetson) and x86_64 (nvidia gpu).

## Examples

### Running pythia from the cli

<details><summary> gst-pylaunch</summary>

You can run familiar pipelines and attach buffer probes from simple python modules.

#### Create Files

<!-- gst-pylaunch probe -->
* Create a file `probe.py` with:

```python

from pythia import objects_per_batch

def gen_detections(batch_meta):
    for frame, detection in objects_per_batch(batch_meta):
        box = detection.rect_params
        yield {
            "frame_num": frame.frame_num,
            "label": detection.obj_label,
            "left": box.left,
            "top": box.top,
            "width": box.width,
            "height": box.height,
            "confidence": detection.confidence,
        }

```

<!-- gst-pylaunch pipeline -->
* Create a file `pipeline.txt` with:

```
uridecodebin
  uri=file://{input}
! identity
  eos-after=30
! nvvideoconvert
! muxer.sink_0
nvstreammux
  name=muxer width=1280 height=720 batch-size=1
! nvinfer
  name=pgie
  config-file-path={pgie-conf}
! nvvideoconvert
! nvdsosd
! nvvideoconvert
! queue
! x264enc
! mp4mux
! filesink location={output}
```

#### Running pythia

<!-- gst-pylaunch console -->
* run the application with:

```console
$ gst-pylaunch \
  -p ./pipeline.txt \
  --pgie-conf=/opt/nvidia/deepstream/deepstream/samples/configs/deepstream-app/config_infer_primary.txt \
  --input=/opt/nvidia/deepstream/deepstream/samples/streams/sample_720p.mp4 \
  --output=/tmp/overlayed.mp4 \
  --probe=probe.py:gen_detections@pgie.src
```

Note the `--pgie-conf`, `--input`, and `--output` cli args were dynamically parsed and
added from the pipeline file. 

This command instructed pythia to do the following:
  1. Load a pipeline from a file located at `./pipeline.txt`, which contains
    `gst-launch`-like syntax with some parameters to be inserted (`input`, `pgie-conf`,
    `output`).
  2. Format the pipelie with `input`, `pgie-conf` and `output` from received parameters.
    (For a more complex syntax, you can install `pythia[jinja]` to use jinja as a
    template backend. See the documentation for more details.)
  4. Setup a buffer probe which internally calls the `gen_detections` method defined in the
    `probe.py` file.
  5. Attach said buffer probe in the `source pad` of the `pgie`-named element of the
    pipeline.
  6. Send incoming metadata to a logger which prints jsonified metadata to console.

#### Check your output

* Check your console to see the incoming detections.
* Want to do something else with the detections? You can choose between several
  backends: logging (stdout <default>, stderr, file available), in-memory (deque),
  kafka, redis, or implement your own streaming connector with the `PYTHIA_STREAM_URI`
  env var. Check the documentation for more details.
  
</details>


### Develop applications based on pythia
<details><summary>python API</summary>


If you want more granular control over the behavior of the application, its signals,
events, and messages, you can instead program an aplication using pythia's API.


#### Create Files

Continuing with the same pipeline as in the previous example,

<!-- api application -->
* Create a file `myscript.py` with:

```python
import json
from kafka import KafkaProducer
from kafka.admin import KafkaAdminClient, NewTopic
from pythia import Application, Gst, objects_per_batch

class App(Application):

    def __init__(self, *a, **kw):
        super().__init__(*a, **kw)
        self.manual_kafka = KafkaProducer(
            bootstrap_servers="kafka:9092"
        )

    def on_message_error(self, *a, **kw):
        err, debug = super().on_message_error(*a, **kw)
        self.manual_kafka.send(
            "app_events",
            json.dumps({ "CONDITION":"ERROR", "ERR": err, "DEBUG": debug}).encode()
        )
        raise RuntimeError("Unhandled pipeline error")

    def on_message_eos(self, bus, message):
        self.manual_kafka.send(
            "app_events",
            json.dumps({ "CONDITION":"EOS", "SENT_BY": str(message.src)}).encode()
        )
        super().on_message_eos(bus, message)

app = App.from_pipeline_file(
    "pipeline.txt",
    params={
      "pgie-conf": "/opt/nvidia/deepstream/deepstream/samples/configs/deepstream-app/config_infer_primary.txt",
      "input": "/opt/nvidia/deepstream/deepstream/samples/streams/sample_720p.mp4",
      "output": "/tmp/overlayed.mp4",
    }
)

@app.probe(
    "pgie",
    pad_direction="src",
    backend_uri="kafka://kafka:9092?stream=raw_detections"
)
def pgie_srcprobe(batch_meta):
    for frame, detection in objects_per_batch(batch_meta):
        frame_num = frame.frame_num
        box = detection.rect_params
        yield {
            "frame_num": frame_num,
            "label": detection.obj_label,
            "left": box.left,
            "top": box.top,
            "width": box.width,
            "height": box.height,
            "confidence": detection.confidence,
        }

@app.probe(
    "muxer",
    pad_direction="src",
)
def source_probe(pad, info):
    app.manual_kafka.send(
        "app_events",
        json.dumps({
            "CONDITION":"STARTED",
            "PAD_CAPS": pad.props.caps.to_string(),
            "PAD_DIRECTION": pad.props.direction,
            "PAD_OFFSET": pad.props.offset,
        }).encode()
    )
    return Gst.PadProbeReturn.REMOVE

if __name__ == "__main__":
    admin = KafkaAdminClient(bootstrap_servers="kafka:9092")
    if "app_events" not in admin.list_topics():
        admin.create_topics(
            new_topics=[
                NewTopic(name="app_events", num_partitions=1,replication_factor=1)
            ],
            validate_only=False,
        )
    app()

```

#### Running pythia

<!-- api console -->
* run the application with:

```console
$ python myscript.py
```

In this mode, you have more control over the application behavior:

1. Subclass application
2. instantiate a custom message handler (a kafka producer in this example)
3. forward error and EOS messages to a custom kafka topic
4. interpolate the pipeline template file with python variables to construct the app
5. use the `@app.probe` decorator as a generator, letting pythia handle the messages
  internally
6. use the `@app.probe` decorator as a probe, handling manually the buffer flow and
  messaging.


Want to do something else while the application is running? you can run the application
with `app(background=True)` instead. See the documentation for details and more
examples.

#### Check your output

* Check the kafka topics to see the incoming detections.

</details>

## Setup

### Prerequisites

* nvidia hardware (either jetson or gpu)
* One of
  - recent docker (with support `--gpus=all`)
  - `nvidia-docker` installed,
  - environment with deepstream 6.1.1 and [these bindings](https://github.com/rmclabs.io/deepstream_python_apps)


### Install

#### non-docker

* `pip install pythiags`

#### docker

* `docker login ghcr.io` (See instructions [here](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry))
* `docker pull ghcr.io/rmclabs-io/pythia` or `ghcr.io/rmclabs-io/pythia-l4t`
* Build your image using `FROM ghcr.io/rmclabs-io/pythia` or
  `FROM ghcr.io/rmclabs-io/pythia-l4t`

NOTE: `latest` tag is deliberately not published.

* run docker with the following flags at least:

  ```console
  $ docker run \
    --gpus=all \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    -e DISPLAY \
    ghcr.io/rmclabs-io/pythia-dev:1.2.1
  ```

Alternatively, you could use `ghcr.io/rmclabs-io/pythia-dev` or
`FROM ghcr.io/rmclabs-io/pythia-l4t-dev`.

## Usage

Note: If running from docker, make sure you've properly configured the container and its
environment, see the
[reference upstream](https://docs.nvidia.com/metropolis/deepstream/dev-guide/text/DS_docker_containers.html#)

For more examples and tutorials, visit the
[examples section](https://dev.rmclabs.io/pythia/docs/examples.html)
of the documentation.

## FAQ

Check out ongoing and future development [here](https://github.com/rmclabs-io/pythiags/projects)

### FAQ / Common Issues

* Q: Package installation fails:

  * A1: upgrade your pip: `pip install --upgrade pip` (Required "pip>=10").
  * A2: Make sure you've installed the build prerequisites, as listed in `reqs/apt.build.list`.

* Q: My application is running slow on Jetson

  * A: Ensure to enable jetson-clocks and maxn (See
    [reference](https://docs.nvidia.com/metropolis/deepstream/dev-guide/text/DS_Performance.html#jetson))
    :

     ```console
     sudo nvpmodel -m 0
     sudo jetson_clocks
     ```

* Q: Program exits with error `Unable to get a Window, abort.`

  * A: Make sure x11 is properly configured. This is common when running through ssh
     sessions. In most of the cases, this just means you need to have the `DISPLAY`
     environment variable correctly set. To list available displays, run the `w`
     command:

     ```bash
     $ w
     09:53:38 up 2 days, 17:26,  1 user,  load average: 0,36, 0,33, 0,23
     USER     TTY      FROM             LOGIN@   IDLE   JCPU   PCPU WHAT
     rmclabs  :0       :0               lun16   ?xdm?   3:06m  0.02s /usr/lib/gdm3/gdm-x-session --run-script /usr/lib/gnome-session/run-systemd-session unity-session.     target
     rmclabs  pts/11   10.100.10.79     09:57    1.00s  0.10s  0.00s w
     ```

     From here, choose a display corresponding to a local connection (`:0` in this case,
     including the colon). Then, export the environment variable and run again your program:

     ```bash
     export DISPLAY=:0
     # run your program here
     ```

* Q: Program exits with error (from docker):

     ```console
     X Error of failed request:  BadShmSeg (invalidshared segment parameter)
       Major opcode of failed request:  150 (XVideo)
       Minor opcode of failed request:  19 ()
       Segment id in failed request:  0x121
       Serial number of failed request:  57
       Current serial number in output stream:  58
     python: ../../src/hb-object-private.hh:154: Type* hb_object_reference(Type*) [with Type = hb_unicode_funcs_t]: Assertion `hb_object_is_valid (obj)' failed.
     ```

  * A: Add `--ipc=host` flag to docker run.

* Q: Python segfaults when several applications are run subsequently:

  * A: It seems to be a race condition produced by the `uridecodebin` element using
    `nvjpegenc` (maybe others?). You can either add a timeout between runs (1 sec seems
    to do it), or change `nvjpegenc` - see `pythia.utils.gst:demote_plugin`.

* Q: Python segfaults when several applications are run subsequently:

  * A: It seems to be a race condition produced by the `uridecodebin` element using
    `nvjpegenc` (maybe others?). You can either add a timeout between runs (1 sec seems
    to do it), or change `nvjpegenc` - see `pythia.utils.gst:demote_plugin`.

* Q: I am unable to build the devcontainer:

  * A: Make sure to update the `devcontainer.json` with a proper `BASE_IMAGE` and
    `DOCKER_GROUP_ID`.

## Contribute

1. fork
2. clone
3. Pull Request from new branch
4. [Optional, recommended]: use provided devcontainer.
5. [Optional, recommended]: run `pre-commit install` to validate commits.
6. add tests, docs, code, scripts, etc
7. [Optional] check code manually, with `./scripts/format`, `./scripts/lint`,
  `./scripts/docs`, `./scripts/test`, etc.
8. Commit using
  [Conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
9. push, wait for ci and/or maintainers feedback
10. repeat 6-8 until success!

For more instructions, visit the
[Developers section](https://rmclabs-io.github.io/pythia-docs/development)
of the documentation.


CHANGELOG.md
## 1.3.0 (2022-10-07)

### Feat

- support for analytics and tracker

## 1.2.4 (2022-09-28)

### Fix

- **tracker**: allow pipelines with nvtracker to run

## 1.2.3 (2022-09-23)

### Fix

- **docker**: include peoplesegnet & trt patch

## 1.2.2 (2022-09-23)

### Fix

- **event_stream/log**: default stream to Path if not in stdin/stdout

## 1.2.1 (2022-09-23)

## 1.2.0 (2022-09-23)

### Feat

- demote plugin util to lower its rank

## 1.1.0 (2022-09-20)

### Fix

- avoid segfault with uridecodebin+nvjpegenc on subsequent runs

### Feat

- **app**: enable loop customization

## 1.0.4 (2022-09-15)

### Fix

- avoid duplicates from workers

## 1.0.3 (2022-09-14)

### Fix

- login to ghcr before publishing docker images from ci pipeline

## 1.0.2 (2022-09-14)

### Fix

- publish docker images from ci pipeline

## 1.0.1 (2022-09-14)

### Fix

- avoid release creation from ci pipeline

## 1.0.0 (2022-09-14)

### Refactor

- rebuild pythia from scratch to improve APIs

## 0.12.1 (2022-02-15)

### Fix

- **recorder**: correct recorder poll max delay

## 0.12.0 (2021-11-05)

### Feat

- **recorder**: enable recordbin, ringbuffer customization (#39)
- **recorder**: enable recordbin, ringbuffer customization

## 0.11.2 (2021-11-05)

### Fix

- **recorder**: allow on_video_finished on multiple calls (#36)

## 0.11.1 (2021-10-06)

### Fix

- **recorder**: allow multiple videorecorder instances in same pipeline (#33)

## 0.11.0 (2021-09-24)

## 0.10.0 (2021-04-22)

### Feat

- **x86**: Add support for GPU devices (non jetson-arm)

## 0.9.0 (2021-04-13)

### Feat

- **deepstream.iterators**: dsanalytics metadata extraction (#24)
- **deepstream.iterators**: deepstream analytics metadata extraction implementation

* update pyds_ext to 1.3, which includes pyds_analytics_metadata
* Additional iterators using new pyds_ext api
  - analytics_per_frame
  - frame_analytics_per_batch
  - analytics_per_object
  - object_analytics_per_frame

## 0.8.1 (2021-03-11)

### Fix

- **processor**: validate processor using real signature form inspect
- **logging**: avoid crash when using kivy logger in debug mode
- **processor**: validate processor using real signature form inspect

### Perf

- **background**: avoid enqueuing on falsy content

## 0.8.0 (2021-02-26)

### Fix

- **app**: pipeline property automativcally defined as get_camera is a requirement
- **app**: allow pipeline play after pausing fails for increased verbosity

### Feat

- **app**: allow settings resolution on pythiags application

## 0.7.1 (2021-02-25)

### Fix

- **app**: background thread init force app crash (#20)
- **app**: ensure application crashes when the background initialization fails

## 0.7.0 (2021-02-25)

### Feat

- **app**: simplify kivy app on_start
- **app**: application-level callback for on_first_frame_out

## 0.6.0 (2021-02-23)

### Feat

- **PythiaGsCli**: timeout running using cli_run (#18)

## 0.5.6 (2021-02-23)

### Fix

- **cli**: improve attachment of processors (#17)

## 0.5.5 (2021-02-22)

### Fix

- **cli**: working kivy and gobject backends
- **cli**: pygst-launch working with Gobject and kivy backends
- **headless**: fix: headless mode startup and closing routine

### Refactor

- tests
- moved processor validation function into utils
- increased api verbosity

## 0.5.4 (2021-02-17)

### Fix

- **pep8names**: ensure class names after pythia refactor are CamelCase (#15)

## 0.5.3 (2021-02-17)

### Fix

- **deps**:  allow pypi publish (#14)

## 0.5.2 (2021-02-16)

### Fix

- **docs**: build documentation instead of isort

## 0.5.1 (2021-02-16)

### Fix

- **bump**: avoid installing project before versionbump (#13)
- **refactor**: moved src to pythiags (#12)
- **deploy**: wron version verification command in deploy pipeline (#8)

### Refactor

- **name**: pythiags (#9)

## 0.5.0 (2021-02-16)

### Refactor

- **configparser**: move functions inside main cls

### Fix

- **ci**: use cr pat correct secret

### Feat

- **tracker**: implementred tracker functionality (#6)

## 0.4.0 (2020-12-22)

## 0.3.2 (2020-12-21)

## 0.3.1 (2020-12-18)

## 0.3.0 (2020-10-30)


src/pythia/video.py
# -*- coding: utf-8 -*-
""""""

import atexit
import os
import sys
from ctypes import Structure
from ctypes import c_int
from ctypes import c_void_p
from ctypes import string_at
from weakref import ref

from kivy.clock import Clock
from kivy.core.camera import CameraBase
from kivy.graphics.texture import Texture
from kivy.properties import StringProperty
from kivy.support import install_gobject_iteration
from kivy.uix.camera import Camera
from kivy.uix.image import Image

from pythia import GLib
from pythia import Gst
from pythia import logger

Gst.init(None)
try:
    install_gobject_iteration()
except AttributeError as exc:
    if "sphinx-build" in sys.argv[0]:
        # NOTE: kivy.clock.Clock is not available <pwoolvett 2021-01-27T18:24:30>
        logger.warning(exc)
    else:
        raise


def parse_launch(gstlaunch_pipeline: str) -> str:
    if "DISPLAY" not in os.environ:
        logger.error("DISPLAY env var not set! This will cause errors")
        raise RuntimeError
    try:
        pipeline = Gst.parse_launch(gstlaunch_pipeline)
    except GLib.GError as exc:
        logger.error(str(exc))
        raise RuntimeError from exc
    if not pipeline:
        raise RuntimeError
    return pipeline


def get_by_name(pipeline: Gst.Pipeline, name: str) -> Gst.Element:
    element = pipeline.get_by_name(name)
    if element:
        return element
    raise ValueError(f"{pipeline}.get_by_name('{name}') returned None")


def get_static_pad(element: Gst.Element, padname: str) -> Gst.Pad:
    pad = element.get_static_pad(padname)
    if pad:
        return pad
    raise ValueError(f"{element}.get_static_pad('{padname}') returned None")


def on_eos(_, __, app):
    logger.info("Gstreamer: End-of-stream\n")
    app.stop()
    return True


def on_error(_, message, app):
    err, debug = message.parse_error()
    logger.error("Gstreamer: %s: %s\n", err, debug)
    app.stop()
    return True


class DeepstreamCamera(CameraBase):
    """Allow arbitrary gst-launch pipeline to be displayed from kivy.

    The following requirements must be met:

    1 The last pipeline element must be the following:

      `appsink name=pythia emit-signals=True caps=video/x-raw,format=RGB`

    2 The pipeline must contain an element named "decoder". This is used
      internally by `kivy.core.camera.camera_gi.CameraGi` to extract the
      texturesize once a successful appsink `pull-sample` is emitter,
      with the following callback:

      .. code-block:: python

         for pad in decoder.srcpads:
             s = pad.get_current_caps().get_structure(0)
             return (s.get_value('width'), s.get_value('height'))

    Additionally, if the pipeline contains an element named "observer".
    It will be used to ectract deepstream metadata

    .. seealso:: kivy.core/camera/camera_gi.CameraGi - this class is
        based on that

    """

    _instances = []

    class MapInfo(Structure):
        _fields_ = [("memory", c_void_p), ("flags", c_int), ("data", c_void_p)]
        # we don't care about the rest

    def __init__(self, pipeline_string, **kwargs):
        self.pipeline_string = pipeline_string
        self._pipeline = None
        self._register_ref()
        kwargs.setdefault("resolution", (-1, -1))
        super().__init__(**kwargs)

    def _register_ref(self):
        cls = type(self)

        def _on_unref(instance):
            if instance in cls._instances:
                cls._instances.remove(instance)

        cls._instances.append(ref(self, _on_unref))

    def init_camera(self):
        if self._pipeline:
            self._pipeline = None
        self._pipeline = parse_launch(self.pipeline_string)

        self._sink = get_by_name(self._pipeline, "pythia")
        get_static_pad(self._sink, "sink").add_probe(
            Gst.PadProbeType.BUFFER, self.on_first_frame_out
        )
        self._sink.connect("new-sample", self._gst_new_sample)

        if self._sink and not self.stopped:
            self.start()

    def on_first_frame_out(self, pad, _):
        struct = pad.get_current_caps().get_structure(0)

        self._texturesize = (
            struct.get_value("width"),
            struct.get_value("height"),
        )
        if self.resolution == (-1, -1):
            self._resolution = self._texturesize

        return Gst.PadProbeReturn.REMOVE

    def _gst_new_sample(self, _):
        sample = self._sink.emit("pull-sample")
        if sample is None:
            return Gst.FlowReturn.ERROR

        self._sample = sample
        Clock.schedule_once(self._update)
        return Gst.FlowReturn.OK

    def start(self):
        super().start()
        self.set_state(Gst.State.PLAYING)

    def stop(self, on_err="raise"):
        super().stop()
        self.set_state(Gst.State.PAUSED, on_err=on_err)

    def unload(self, on_err="raise"):
        self.set_state(Gst.State.NULL, on_err=on_err)

    def set_state(self, state: Gst.State, on_err="raise"):
        try:
            result = self._pipeline.set_state(state)
            if result == Gst.StateChangeReturn.FAILURE:
                raise RuntimeError
        except (AttributeError, RuntimeError) as err:
            if on_err == "warn":
                return logger.error(err)
            if on_err == "raise":
                raise

    def _update(self, dt):
        sample, self._sample = self._sample, None
        if sample is None:
            return
        if self._texture is None:
            self._texture = Texture.create(
                size=self._texturesize, colorfmt="rgb"
            )
            self._texture.flip_vertical()
            self.dispatch("on_load")

        # decode sample
        # read the data from the buffer memory
        try:
            buf = sample.get_buffer()
            _, mapinfo = buf.map(Gst.MapFlags.READ)

            # We cannot get the data out of mapinfo, using Gst 1.0.6 + Gi 3.8.0
            # related bug report:
            # https://bugzilla.gnome.org/show_bug.cgi?id=6t8663
            # ie: mapinfo.data is normally a char*, but here, we have an int
            # So right now, we use ctypes instead to read the mapinfo ourself.
            addr = mapinfo.__hash__()
            c_mapinfo = self.MapInfo.from_address(addr)

            # now get the memory
            self._buffer = string_at(c_mapinfo.data, mapinfo.size)
            self._copy_to_gpu()
        finally:
            if mapinfo is not None:
                buf.unmap(mapinfo)

    @classmethod
    def camera_gi_clean(cls):
        """Forcing the stop/unload of all remaining videos before  exiting the
        python process.

        If we leave the python process with some video running, we can
        hit a segfault.

        """
        for weakcamera in cls._instances:
            camera = weakcamera()
            if isinstance(camera, DeepstreamCamera):
                camera.stop(on_err="ignore")
                camera.unload(on_err="ignore")


class GSCameraWidget(Camera):
    """CameraWidget which receives arbitrary Gstreamer pipelines as parameter.

    TODO: sync resolution param with the output of the pipeline_string

    """

    pipeline_string = StringProperty()
    """Pipeline_string to build Gstreamer Pipeline."""

    def __init__(self, **kwargs):
        """Override to `fbind` `pipeline_string` as well."""
        self._camera = None
        super(Image, self).__init__(
            **kwargs
        )  # Image's init to avoid Camera to avoid Camera form calling `on_index` at the end of its constructor
        if self.index == -1:
            self.index = 0
        on_index = self._on_index
        fbind = self.fbind
        fbind("index", on_index)
        # fbind("resolution", on_index)
        fbind("pipeline_string", on_index)
        on_index()

    def _on_index(self, *largs):
        """Override super to enforce `DeepstreamCamera`"""
        self._camera = None
        if not self.pipeline_string:
            return
        self._camera = DeepstreamCamera(
            index=self.index,
            # resolution=self.resolution,
            pipeline_string=self.pipeline_string,
            stopped=True,
        )
        self._camera.bind(on_load=self._camera_loaded)
        if self.play:
            self._camera.start()
            self._camera.bind(on_texture=self.on_tex)


atexit.register(DeepstreamCamera.camera_gi_clean)
